weaver/src/main/java/org/aspectj/weaver/reflect/Java15AnnotationFinder.java

```diff
@@ -57,12 +65,29 @@ public Java15AnnotationFinder() {
 	}
 
 	public void setClassLoader(ClassLoader aLoader) {
-		this.classLoaderRef = new BcelWeakClassLoaderReference(aLoader);
+		//Set class loader ref
+		if (useSingleInstances && staticClassLoaderRef == null)
+			staticClassLoaderRef = new BcelWeakClassLoaderReference(aLoader);
+		else
+			this.classLoaderRef = new BcelWeakClassLoaderReference(aLoader);
+
+		//Set repository
 		if (useCachingClassLoaderRepository) {
-			this.bcelRepository = new ClassLoaderRepository(classLoaderRef);
-		} else {
-			this.bcelRepository = new NonCachingClassLoaderRepository(classLoaderRef);
+			if (useSingleInstances && staticBcelRepository == null)
+				staticBcelRepository = new ClassLoaderRepository(getClassLoader());
+			else
+				this.bcelRepository = new ClassLoaderRepository(getClassLoader());
 		}
+		else {
+			if (useSingleInstances && staticBcelRepository == null)
+				staticBcelRepository = new NonCachingClassLoaderRepository(getClassLoader());
+			else
+				this.bcelRepository = new NonCachingClassLoaderRepository(getClassLoader());
```

Comment:
For 100% backward compatibility, you should use `this.classLoaderRef = new BcelWeakClassLoaderReference(aLoader); like in the original code`, because `getClassLoader()` returns `classLoaderRef.getClassLoader()` and there is no guarantee that this is always the same as the method parameter `aLoader`.